lib/ConstExtract/ConstExtract.cpp

```diff
@@ -207,33 +248,52 @@ gatherConstValuesForPrimary(const std::unordered_set<std::string> &Protocols,
   return Result;
 }
 
-std::string toString(const CompileTimeValue *Value) {
-  switch (Value->getKind()) {
-    case CompileTimeValue::RawLiteral:
-      return cast<RawLiteralValue>(Value)->getValue();
-    case CompileTimeValue::InitCall:
-      // TODO
-    case CompileTimeValue::Builder:
-      // TODO
-    case CompileTimeValue::Dictionary:
-      // TODO
-    case CompileTimeValue::Runtime:
-      return "Unknown";
+void writeProperty(llvm::json::OStream &JSON,
+                   const ConstValueTypePropertyInfo PropertyInfo) {
+  const auto *PropertyDecl = PropertyInfo.VarDecl;
+  JSON.attribute("label", PropertyDecl->getName().str().str());
+  JSON.attribute("type",
+                 toFullyQualifiedTypeNameString(PropertyDecl->getType()));
+  JSON.attribute("isStatic", PropertyDecl->isStatic() ? "true" : "false");
+  JSON.attribute("isComputed", !PropertyDecl->hasStorage() ? "true" : "false");
+
+  auto value = PropertyInfo.Value.get();
+  switch (value->getKind()) {
+  case CompileTimeValue::ValueKind::RawLiteral:
+    JSON.attribute("valueKind", "RawLiteral");
+    JSON.attribute("value", cast<RawLiteralValue>(value)->getValue());
+    break;
+
+  case CompileTimeValue::ValueKind::InitCall: {
+    auto initCallValue = cast<InitCallValue>(value);
+
+    JSON.attribute("valueKind", "InitCall");
+    JSON.attributeObject("value", [&]() {
+      JSON.attribute("type", initCallValue->getName());
+      JSON.attributeArray("arguments", [&] {
+        for (auto FP : initCallValue->getFunctionParameters()) {
+          JSON.object([&] {
+            JSON.attribute("label", FP.Label);
+            JSON.attribute("type", toFullyQualifiedTypeNameString(FP.Type));
+            JSON.attribute("value", FP.Value);
+          });
+        }
+      });
+    });
+    break;
   }
-}
 
-std::string toString(CompileTimeValue::ValueKind Kind) {
-  switch (Kind) {
-    case CompileTimeValue::ValueKind::RawLiteral:
-      return "RawLiteral";
-    case CompileTimeValue::ValueKind::InitCall:
-      return "InitCall";
-    case CompileTimeValue::ValueKind::Builder:
-      return "Builder";
-    case CompileTimeValue::ValueKind::Dictionary:
-      return "Dictionary";
-    case CompileTimeValue::ValueKind::Runtime:
-      return "Runtime";
```

Comment:
Nice, I like how you inlined this switch in the single call site, and reduced calls to `getKind()`.

Reply:
Thanks. 

I've had some thoughts on how we might iterate on the JSON rendering in follow-up changes. Similar to your comment about storing extracted const recursively, I imagine we may want to extract out simple JSON property writers that can handle the recursive values.

Reply:
Excellent idea. It sounds reasonable to develop recursive parsing and JSON output in the same subsequent change.